### PR TITLE
Update exception to provide more info + Ignore cleared IDs for unlocks

### PR DESCRIPTION
**Pull request by ashrayjain**
@SerialVelocity I think we might have encountered that error due to an existing cleared lock? This might happen between consecutive unlock attempts right? Regardless, I have ignored the lockId check if the existing value is cleared.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1069)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 1]
```diff
@@ -306,9 +306,13 @@ private boolean trySchemaMutationUnlockOnce(long perOperationNodeId) {
         try {
             result = clientPool.runWithRetry(client -> {
                 Column existingColumn = queryExistingLockColumn(client);
+                long existingLockId = getLockIdFromColumn(existingColumn);
 
-                Preconditions.checkState(isValidColumnForLockId(existingColumn, perOperationNodeId),
-                        "Trying to unlock unowned lock");
+                if (existingLockId != GLOBAL_DDL_LOCK_CLEARED_ID) {
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: This seems odd. Why aren't we doing a return instead? Make this runWithRetry return whether it was successful or not instead.

Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: And is there a way to test this case?

Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Sure, we could update this so that runWithRetry returns a boolean. Not sure how we will test this though, since I believe we don't have a way of just trying an unlock. The only API we have is `runWithLock` right?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 2]
```diff
@@ -302,26 +302,31 @@ private void schemaMutationUnlock(long perOperationNodeId) {
...
+                Preconditions.checkState(existingLockId == perOperationNodeId,
+                        String.format("Trying to unlock unowned lock. Expected [%d], but got [%d]",
+                                perOperationNodeId, existingLockId));
+
                 List<Column> ourExpectedLock = ImmutableList.of(existingColumn);
                 Column clearedLock = lockColumnWithValue(GLOBAL_DDL_LOCK_CLEARED_VALUE);
-                return writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                CASResult casResult = writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                if (casResult.isSuccess()) {
+                    log.info("Successfully released schema mutation lock.");
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: We should mention the lock ID. If we add this in, we should make sure there's a similar `log.info` for when we take out the lock.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 3]
```diff
@@ -302,26 +302,31 @@ private void schemaMutationUnlock(long perOperationNodeId) {
     }
 
     private boolean trySchemaMutationUnlockOnce(long perOperationNodeId) {
-        CASResult result;
+        boolean result;
         try {
             result = clientPool.runWithRetry(client -> {
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Don't store this in `result`. Just `return`

Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: ok

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 4]
```diff
@@ -217,7 +217,7 @@ private long waitForSchemaMutationLock() {
                 }
 
                 // we won the lock!
-                log.info("Successfully acquired schema mutation lock.");
+                log.info(String.format("Successfully acquired schema mutation lock with id [%d]", perOperationNodeId));
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: log.info is a vararg. Don't use `String.format`

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 5]
```diff
@@ -302,22 +302,29 @@ private void schemaMutationUnlock(long perOperationNodeId) {
...
                 Column existingColumn = queryExistingLockColumn(client);
+                long existingLockId = getLockIdFromColumn(existingColumn);
+                if (existingLockId == GLOBAL_DDL_LOCK_CLEARED_ID) {
+                    return true;
+                }
 
-                Preconditions.checkState(isValidColumnForLockId(existingColumn, perOperationNodeId),
-                        "Trying to unlock unowned lock");
+                Preconditions.checkState(existingLockId == perOperationNodeId,
+                        String.format("Trying to unlock unowned lock. Expected [%d], but got [%d]",
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Same here.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 6]
```diff
@@ -302,22 +302,29 @@ private void schemaMutationUnlock(long perOperationNodeId) {
...
+                                perOperationNodeId, existingLockId));
 
                 List<Column> ourExpectedLock = ImmutableList.of(existingColumn);
                 Column clearedLock = lockColumnWithValue(GLOBAL_DDL_LOCK_CLEARED_VALUE);
-                return writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                CASResult casResult = writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                if (casResult.isSuccess()) {
+                    log.info(String.format("Successfully released schema mutation lock with id [%d]", existingLockId));
```
Comment (SerialVelocity) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: And here.

**Comment by ashrayjain** [comment]
@SerialVelocity anything else here? If not, let's merge this.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 7]
```diff
@@ -302,22 +302,29 @@ private void schemaMutationUnlock(long perOperationNodeId) {
...
+                Preconditions.checkState(existingLockId == perOperationNodeId,
+                        "Trying to unlock unowned lock. Expected [%s], but got [%s]",
+                        perOperationNodeId, existingLockId);
 
                 List<Column> ourExpectedLock = ImmutableList.of(existingColumn);
                 Column clearedLock = lockColumnWithValue(GLOBAL_DDL_LOCK_CLEARED_VALUE);
-                return writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                CASResult casResult = writeDdlLockWithCas(client, ourExpectedLock, clearedLock);
+                if (casResult.isSuccess()) {
+                    log.info("Successfully released schema mutation lock with id [{}]", existingLockId);
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: Can we log this in `schemaMutationUnlock` under `if(unlockDone)`?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java [hunk 8]
```diff
@@ -302,22 +302,29 @@ private void schemaMutationUnlock(long perOperationNodeId) {
...
                 Column existingColumn = queryExistingLockColumn(client);
+                long existingLockId = getLockIdFromColumn(existingColumn);
+                if (existingLockId == GLOBAL_DDL_LOCK_CLEARED_ID) {
+                    return true;
+                }
 
-                Preconditions.checkState(isValidColumnForLockId(existingColumn, perOperationNodeId),
-                        "Trying to unlock unowned lock");
+                Preconditions.checkState(existingLockId == perOperationNodeId,
+                        "Trying to unlock unowned lock. Expected [%s], but got [%s]",
```
Comment (gbrova) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: It's not clear from this message which of {expected, got} is the one we own. Maybe rephrase as "Trying to unlock lockId=%s but a different lock with lockId=%s is taken out."

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java (hunk 7 above)
Comment (ashrayjain) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLock.java]: We could but the log looks weird that way, since we log the "acquired message" inside `runWithRetry`. There is additional text wrapped around the log when logging from inside `runWithRetry` so this was needed for consistency.
